### PR TITLE
RuboCopBear add configs 

### PR DESCRIPTION
**Pull request by abhsag24**

### CODE REVIEW

File: tests/ruby/RuboCopBearTest.py [hunk 1]
```diff
@@ -17,3 +19,24 @@
...
+def goodName
+  some content
+end
+""".splitlines(keepends=True)
+
+snake_case_file = """
+def bad_name
+  some content
+end
+""".splitlines(keepends=True)
```
Comment (sils1297) [inline tests/ruby/RuboCopBearTest.py]: can you do the testcase with the previously defined files? Less duplication and you get proof that the config is changing behaviour, not the different content.

Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: aahh sure will do it now,

File: bears/ruby/RuboCopBear.py [hunk 2]
```diff
@@ -25,11 +25,185 @@ class RuboCopBear:
     def create_arguments(filename, file, config_file, rubocop_config: str=""):
         # Need both stdin and filename. Explained in this comment:
         # https://github.com/bbatsov/rubocop/pull/2146#issuecomment-131403694
-        args = (filename, '--stdin', '--format=json')
+        args = (filename, '--stdin', '--format=json', )
```
Comment (sils1297) [inline bears/ruby/RuboCopBear.py]: unneeded and unrelated change

File: bears/ruby/RuboCopBear.py [hunk 3]
```diff
@@ -25,11 +25,185 @@ class RuboCopBear:
...
+                        allow_unused_block_keyword_arguments: bool=False,
+                        ignore_unused_method_args_if_empty: bool=True,
+                        allow_unused_method_keyword_args: bool=False):
+        """
+        Not all settings added.
+        Notable settings missing: Rails settings:
+
+        :param access_modifier_indentation:
+            Indent private/protected/public as deep as method definitions
+            options: 'indent', 'outdent'.
```
Comment (sils1297) [inline bears/ruby/RuboCopBear.py]: what's indent and outdent? Doesn't get fully clear to me.

File: bears/ruby/RuboCopBear.py [hunk 4]
```diff
@@ -25,11 +25,185 @@ class RuboCopBear:
...
+        :param preffered_alias:
+            Alias by ``alias`` or ``alias_method``.
```
Comment (sils1297) [inline bears/ruby/RuboCopBear.py]: I don't understand what this is for.

File: tests/ruby/RuboCopBearTest.py [hunk 5]
```diff
@@ -40,3 +41,33 @@ def bad_name
...
+'using single-quotes'
+""".splitlines(keepends=True)
+good_file = """
+"using double-quotes"
+""".splitlines(keepends=True)
+RuboCopBearSettingsTest = verify_local_bear(
+                              RuboCopBear,
+                              valid_files=(good_file,),
+                              invalid_files=(bad_file,),
+                              settings={"string_literals": "double_quotes"})
```
Comment (sils1297) [inline tests/ruby/RuboCopBearTest.py]: possibly tests can be similified, less sample files are usually better

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/425/commits/e672f3ebf218e0b8ac3c22b842c463aacae6f843

**Comment by sils1297** [comment]
https://github.com/coala-analyzer/coala-bears/pull/425/commits/968b09056445713058b9d3095a8077855c670873 needs work

**Comment by sils1297** [comment]
https://github.com/coala-analyzer/coala-bears/pull/425/commits/4c8be27d09081e14f5b7259f52209eba09039706 as well needs work

File: bears/ruby/RuboCopBear.py (hunk 4 above)
Comment (abhsag24) [inline bears/ruby/RuboCopBear.py]: a different name for a function so you can do it by alias, or alias_method, and ruby projects generally pick one and stay consistent with it, i'll clarify in the update :P

**Comment by sils1297** [comment]
@abhsag24 need something? This looks quickly doable. (And is awesome :))

**Comment by abhsag24** [comment]
@sals1275  doing it now

File: tests/ruby/RuboCopBearTest.py (hunk 5 above)
Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: what do you suggest? i can drop this tests actually since i tested for settings earlier, but i didn't feel good by testing just one setting, or i can mix thee two settings in the earlier example?

Comment (AbdealiJK) [inline tests/ruby/RuboCopBearTest.py]: I tend to disagree with @sils1297 
Its better to have more sample files each with only 1 possible Result. When using 1 big test file, things get messy and it is not a "unit"test anymore.

Possibly you could simplify your code for easier reading. Like `bad_file = ("'using single-quotes'",)` instead of the whole `"""` thing

Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: thanks  :D

File: tests/ruby/RuboCopBearTest.py [hunk 6]
```diff
@@ -1,19 +1,58 @@
+import os
+
 from bears.ruby.RuboCopBear import RuboCopBear
 from tests.LocalBearTestHelper import verify_local_bear
 
-good_file = """def bad_name
+good_file = """def good_name
```
Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: ahh i guess this change belonged in a different commit? but i had to change this test anyway, so i figured i could keep it here?

File: tests/ruby/RuboCopBearTest.py (hunk 5 above)
Comment (sils1297) [inline tests/ruby/RuboCopBearTest.py]: @AbdealiJK if you use the same sample files you basically prove that your configuration changes the results and it's not possible that your other sample files accidentally match

File: tests/ruby/RuboCopBearTest.py (hunk 1 above)
Comment (sils1297) [inline tests/ruby/RuboCopBearTest.py]: still not done?

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala-bears/pull/425/commits/713de13e4761de2290863beea6460f6cad91cd7b

File: bears/ruby/RuboCopBear.py (hunk 2 above)
Comment (sils1297) [inline bears/ruby/RuboCopBear.py]: ping?

File: tests/ruby/RuboCopBearTest.py (hunk 1 above)
Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: they were in the different commit fixing it :(

File: tests/ruby/RuboCopBearTest.py [hunk 7]
```diff
@@ -1,19 +1,32 @@
...
   test if something
 end
 """.splitlines(keepends=True)
 
 bad_file = """def badName
-  if something
-    test
-    end
```
Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: changed this since this caused more than one errors in the `bad_file`

**Comment by abhsag24** [comment]
CI passes after rebuild, i dont know what problem codecov has , as coverage is 100%

**Comment by AbdealiJK** [comment]
Adding comments on https://github.com/coala-analyzer/coala-bears/commit/a9c13588ade434c91d488bd268f90bb421d87edc

**Comment by AbdealiJK** [comment]
@sils1297 Im getting a little confused tracking all the settings. You've added a bunch of stuff in your commit at https://github.com/coala-analyzer/coala-bears/pull/434 and as they're both Ruby there are bound to be common configs. And it is difficult to keep track of all the other configs that are in other bears too. I think we need a better system.

Some concerns in settings:
 - `max_class_length` - A similar setting is being created by sils in that commit.
 - `max_line_length`, `cyclomatic_complexity` - Popular variable names, and the description for this is repeated in each bear. Redundancy :(

**Comment by sils1297** [comment]
I totally agree @AbdealiJK , file an issue?

**Comment by abhsag24** [comment]
@fneu need review here :D

File: bears/ruby/RuboCopBear.py [hunk 8]
```diff
@@ -28,8 +28,186 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+            with one level of indentation relative to the start of the line
+            with the method call.
+        :param class_check:
+            How to check type of class.
+            options: 'is_a?', 'kind_of?'
+        :param comment_keywords:
+            Checks formatting of special comments.
+        :param min_if_unless_lines:
+            Defines the number of lines of the body of an if / unless
+            needs to have to trigger this cop.
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: this is still mixed up somehow, isn't it?

maybe "Defines the number of lines the body of an if/unless needs to trigger this cop."

File: bears/ruby/RuboCopBear.py [hunk 9]
```diff
@@ -28,8 +28,186 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+
+            The `with_first_parameter` style aligns the following lines along
+            the same column as the first parameter.
+
+            The `with_fixed_indentation` style aligns the following lines
+            with one level of indentation relative to the start of the line
+            with the method call.
+        :param class_check:
+            How to check type of class.
+            options: 'is_a?', 'kind_of?'
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: nitpick, but [codestyle](https://github.com/coala-analyzer/coala/blob/master/docs/Getting_Involved/Codestyle.rst) says to putt a full stop at the end of every message in the docstring. You're kinda inconsistent here, compare lines 113, 116, 119...

File: tests/ruby/RuboCopBearTest.py [hunk 10]
```diff
@@ -30,3 +30,30 @@
...
+""".splitlines(keepends=False)
+
+RuboCopBearSettingsTest = verify_local_bear(
+                              RuboCopBear,
+                              valid_files=(good_file,),
+                              invalid_files=(bad_file,),
+                              settings={"indentation_width": 1})
+
+bad_file = ("'using single quote'",)
+good_file = ('"using double quote"',)
```
Comment (fneu) [inline tests/ruby/RuboCopBearTest.py]: while I agree that this looks better in this case, it is now inconsistent with lines 6, 11, 35, and 41. 

I don't have a strong preference, but maybe decide on a uniform look?

File: bears/ruby/RuboCopBear.py [hunk 11]
```diff
@@ -28,8 +28,186 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+        :param count_keyword_args:
+            Count keyword args while counting all arguments?
+        :param ignore_unused_block_args_if_empty:
+            Ignore unused block arguments if block is empty.
+        :param allow_unused_block_keyword_arguments:
+            Allow unused block keyword arguments.
+        :param ignore_unused_method_args_if_empty:
+            Allows unused method argument if method is empty.
+        :param allow_unused_method_keyword_args:
+            Allows unused keyword arguments in a method.
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: quick thought: this is a pretty long list of stuff and the order doesn't seem clear to me. The following options are more or less alphabetically sorted. would it be possible/ make any sense to sort the params here (and probably in the function on top) or is there a reason for this particular order?

File: bears/ruby/RuboCopBear.py [hunk 12]
```diff
@@ -28,8 +28,186 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+                  'Lint/UnusedBlockArgument':
+                      {'IgnoreEmptyBlocks': ignore_unused_block_args_if_empty,
+                       'AllowUnusedKeywordArguments':
+                           allow_unused_block_keyword_arguments},
+                  'Lint/UnusedMethodArgument':
+                      {'AllowUnusedKeywordArguments':
+                           allow_unused_method_keyword_args,
+                       'IgnoreEmptyMethods':
+                           ignore_unused_method_args_if_empty},
+                  'Metrics/CyclomaticComplexity':
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: style... style... metrics... metrics... lint... lint... metrics. huh?

maybe this can also be sorted in with the other metrics?

**Comment by fneu** [comment]
unack https://github.com/coala-analyzer/coala-bears/pull/425/commits/e6042a86e860ed543c843f9cab4751719f62edac

**Comment by fneu** [comment]
@abhsag24 only nitpicks, really

File: bears/ruby/RuboCopBear.py (hunk 12 above)
Comment (abhsag24) [inline bears/ruby/RuboCopBear.py]: ahh yes i kinda nissed it at first, i didnt care about the order then, i'll fix it in the next iteration

File: tests/ruby/RuboCopBearTest.py (hunk 10 above)
Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: it's inconsistent but it's different too right? i mean it's nearly not a code sample in the first place, like rest of the others

File: tests/ruby/RuboCopBearTest.py (hunk 5 above)
Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: @sils1297  i guess i'll change the example to the snake_case thing that ways we'll have simpler and also non-new test cases

File: bears/ruby/RuboCopBear.py (hunk 11 above)
Comment (abhsag24) [inline bears/ruby/RuboCopBear.py]: the order is based on the order i use in the dictionary at 155 except the one you commented on, i'll fix that up and then the order's will match :)

File: bears/ruby/RuboCopBear.py (hunk 8 above)
Comment (abhsag24) [inline bears/ruby/RuboCopBear.py]: they had that in their documentation i'll play around with it, and see if i can come up with a better documentation

**Comment by abhsag24** [comment]
@fneu ready for another round of reviews? :P

File: bears/ruby/RuboCopBear.py [hunk 13]
```diff
@@ -21,10 +22,191 @@ class RuboCopBear:
...
+            The `with_fixed_indentation` style aligns the following lines
+            with one level of indentation relative to the start of the line
+            with the method call.
+        :param class_check:
+            How to check type of class.
+            options: 'is_a?', 'kind_of?'.
+        :param comment_keywords:
+            Checks formatting of special comments.
+        :param min_if_unless_guard:
+            the number of lines that are tolerable within an if/unless block,
```
Comment (SanketDG) [inline bears/ruby/RuboCopBear.py]: The number...

File: requirements.txt [hunk 14]
```diff
@@ -27,3 +27,4 @@ vim-vint~=0.3.6
 # Use >=3.3 once nltk-3.3 is released.
 nltk~=3.1.0
 appdirs~=1.4.0
+pyyaml~=3.11
```
Comment (fneu) [inline requirements.txt]: Shouldn't this be another commit? Might want to split that.

File: tests/ruby/RuboCopBearTest.py [hunk 15]
```diff
@@ -30,3 +30,28 @@
...
+                              RuboCopBear,
+                              valid_files=(another_good_file,),
+                              invalid_files=(another_bad_file,),
+                              settings={"indentation_width": 1})
+
+RuboCopBearSettingsTest = verify_local_bear(
+                              RuboCopBear,
+                              valid_files=(bad_file,),
+                              invalid_files=(good_file,),
+                              settings={"method_name": "camelCase"})
```
Comment (fneu) [inline tests/ruby/RuboCopBearTest.py]: the other test is gone now?

File: bears/ruby/RuboCopBear.py [hunk 16]
```diff
@@ -21,10 +22,191 @@ class RuboCopBear:
...
+            the same column as the first parameter.
+
+            The `with_fixed_indentation` style aligns the following lines
+            with one level of indentation relative to the start of the line
+            with the method call.
+        :param class_check:
+            How to check type of class.
+            options: 'is_a?', 'kind_of?'.
+        :param comment_keywords:
+            Checks formatting of special comments.
```
Comment (SanketDG) [inline bears/ruby/RuboCopBear.py]: Checks formatting of special comments based on keywords like TODO, FIXME etc

File: bears/ruby/RuboCopBear.py [hunk 17]
```diff
@@ -21,10 +22,191 @@ class RuboCopBear:
...
+        :param min_if_unless_guard:
+            the number of lines that are tolerable within an if/unless block,
+            more than these lines call for the usage of a guard clause.
+        :param indentation_width:
+            No. of spaces to indent.
+        :param method_name:
```
Comment (SanketDG) [inline bears/ruby/RuboCopBear.py]: `method_name_case`?

`method_name` just seems like related with rules specific to naming methods. Its better to be explicit.

I guess others like `variable_name` etc should also be changed.

File: bears/ruby/RuboCopBear.py [hunk 18]
```diff
@@ -21,10 +22,191 @@ class RuboCopBear:
...
+        :param max_module_length:
+            Max lines in a module.
+        :param module_length_count_comments:
+            Count comments while calculating module length?
+        :param cyclomatic_complexity:
+            Cyclomatic Complexity of the file.
+        :param max_line_length:
+            Max length of a line.
+        :param line_length_allow_here_doc:
+            Allow here-doc lines to be more than Max in length.
```
Comment (SanketDG) [inline bears/ruby/RuboCopBear.py]: Allow here-doc lines to be more than the max line length.

File: bears/ruby/RuboCopBear.py [hunk 19]
```diff
@@ -21,10 +22,191 @@ class RuboCopBear:
...
+        :param line_length_allow_uri:
+            To make it possible to copy or click on URIs in the code,
+            we allow ignore long lines containing a URI to be longer than Max.
```
Comment (SanketDG) [inline bears/ruby/RuboCopBear.py]: to be longer than max line length

File: tests/ruby/RuboCopBearTest.py (hunk 15 above)
Comment (abhsag24) [inline tests/ruby/RuboCopBearTest.py]: as @sils1297 suggested that we should test new ways with the same settings since that proves that the method works, and it wasn't just luck, so i changed that test-case to align with the `case

File: requirements.txt (hunk 14 above)
Comment (abhsag24) [inline requirements.txt]: yeah i'll make that a different commit

File: bears/ruby/RuboCopBear.py (hunk 17 above)
Comment (abhsag24) [inline bears/ruby/RuboCopBear.py]: yeah i remember having that thought dont know why i never did that :P

**Comment by abhsag24** [comment]
@fneu reviews needed again

**Comment by fneu** [comment]
Awesome, unfortunately I will be busy this afternoon so it will be a couple 
of hours again.

**Comment by sils1297** [comment]
@fneu review please

File: bears/ruby/RuboCopBear.py [hunk 20]
```diff
@@ -28,8 +28,188 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+        Notable settings missing: Rails settings.
+
+        :param access_modifier_indentation:
+            Indent private/protected/public as deep as method definitions
+            options:
+                ``indent`` :  Indent modifiers like class members.
+                ``outdent`` : Indent modifiers one level less than
+                              class members.
+        :param preferred_alias:
+            Which method to use for aliasing in ruby?
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: The question mark puts me off here...

I'd go with just "Method to use for aliasing in ruby" since that is what the param defines. 

Or is this documented like that in rubocop?

File: bears/ruby/RuboCopBear.py [hunk 21]
```diff
@@ -28,8 +28,188 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+            Case of a variable's name.
+            options: 'snake_case', 'camelCase'.
+        :param max_class_length:
+            Max lines in a class.
+        :param class_length_count_comments:
+            Count comments while calculating class length?
+        :param max_module_length:
+            Max lines in a module.
+        :param module_length_count_comments:
+            Count comments while calculating module length?
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: Similar thing with the question mark. I get that you want to tell me that this is a bool value, but it really looks like you were unsure about the option :smile: 

I'd go with:
"Whether or not to count comments..." or something similar.

Again I would consider this fine if it is worded like this in Rubocop.

File: bears/ruby/RuboCopBear.py [hunk 22]
```diff
@@ -28,8 +28,188 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+        :param line_length_allow_here_doc:
+            Allow here-doc lines to be more than the max line length.
+        :param line_length_allow_uri:
+            To make it possible to copy or click on URIs in the code,
+            we allow ignore long lines containing a URI to be longer than max
+            line length.
+        :param max_method_length:
+            Max number of lines in a method.
+        :param method_count_comments:
+            Count full line comments?
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: same thing

File: bears/ruby/RuboCopBear.py [hunk 23]
```diff
@@ -28,8 +28,188 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+        :param max_parameters:
+            Max number of parameters in parameter list.
+        :param count_keyword_args:
+            Count keyword args while counting all arguments?
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: same

**Comment by fneu** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/425/commits/c6dced119e78ce2e335708b7065f2007c5d6931e, https://github.com/coala-analyzer/coala-bears/pull/425/commits/af2758d471d87feb0c89f0369ddf9577fc7ccdad, https://github.com/coala-analyzer/coala-bears/pull/425/commits/381ac7b202dd2be74ca8f361fff647e0e424cdce

**Comment by fneu** [comment]
unack https://github.com/coala-analyzer/coala-bears/pull/425/commits/60562932424ef5e8ff65ad62760f4c1079d4e0e8

a few questions remain :)

File: bears/ruby/RuboCopBear.py (hunk 20 above)
Comment (abhsag24) [inline bears/ruby/RuboCopBear.py]: most of the stuff with question marks was documented like that, but i think we need  to make it better anyways , here's the rubocops [defautlt.yml](https://github.com/bbatsov/rubocop/blob/master/config/default.yml)

**Comment by abhsag24** [comment]
@fneu made the required changes need review again :)

File: bears/ruby/RuboCopBear.py [hunk 24]
```diff
@@ -28,8 +28,190 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+                        ignore_unused_method_args_if_empty: bool=True,
+                        allow_unused_method_keyword_args: bool=False):
+        """
+        Not all settings added.
+        Notable settings missing: Rails settings.
+
+        :param access_modifier_indentation:
+            Indent private/protected/public as deep as method definitions
+            options:
+                ``indent`` :  Indent modifiers like class members.
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: why not single quotes...

File: bears/ruby/RuboCopBear.py [hunk 25]
```diff
@@ -28,8 +28,190 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+                ``outdent`` : Indent modifiers one level less than
+                              class members.
+        :param preferred_alias:
+            Which method to use for aliasing in ruby.
+            options : ``alias`` , ``alias_method``.
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: same

File: bears/ruby/RuboCopBear.py [hunk 26]
```diff
@@ -28,8 +28,190 @@ def create_arguments(filename, file, config_file, rubocop_config: str=""):
...
+        :param align_hash_rocket_by:
+            Alignment of entries using hash rocket as separator.
+        :param align_colon_by:
+            Alignment of entries using colon as separator.
+        :param inspect_last_argument_hash:
+            Select whether hashes that are the last argument in a method call
+            should be inspected.
+            options: 'always_inspect', 'always_ignore', 'ignore_implicit',
```
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: ... like here?

**Comment by fneu** [comment]
unack https://github.com/coala-analyzer/coala-bears/pull/425/commits/d553aaec0c9ee187db7b5c4daa02f456283fd438

**Comment by fneu** [comment]
@abhsag24 just noticed one more thing.

looks good besides

File: bears/ruby/RuboCopBear.py (hunk 24 above)
Comment (abhsag24) [inline bears/ruby/RuboCopBear.py]: it's rst right so it will be double backticks would be highlighted, while single quote will remain light that

**Comment by fneu** [comment]
what's wrong with codecov?

File: bears/ruby/RuboCopBear.py (hunk 24 above)
Comment (fneu) [inline bears/ruby/RuboCopBear.py]: uh I didn't know that ^^

Sure in this case convert them all to backticks

**Comment by abhsag24** [comment]
AppVeyor failes sometimes due to time out and codecov fails whenever there is a CI failure

**Comment by abhsag24** [comment]
@fneu review :P hopefully this one does it :D

**Comment by fneu** [comment]
I guessed so (codecov) but it doesn't make a lot of sense. "Uh, something failed so there is probably something wrong. Let's make it harder to find out what happened."

Commit seems fine :)

**Comment by fneu** [comment]
ack https://github.com/coala-analyzer/coala-bears/commit/deef759a7b82c2376a933c595310490019220b95

**Comment by fneu** [comment]
@rultor merge
